picoscope/picobase.py

```diff
@@ -881,9 +881,16 @@ def close(self):
         might take some time.
 
         """
-        if self.handle is not None:
-            self._lowLevelCloseUnit()
-            self.handle = None
+        try:
+            if self.handle is not None:
+                self._lowLevelCloseUnit()
+                self.handle = None
+        except AttributeError:
+        # _lowLevelCloseUnit raise: libps5000a.so not found if not connected
+            pass
+        except OSError:
+        # self.handle doesn't exist when not connected
```

Comment:
doesn't this raise an attribute error?

Reply:
Oh, sorry, I switched the comments ...

Reply:
what does "handle does not exist" mean? isn't the handle `None` in that case? in which case the if statement takes care of it.

Reply:
That's what I thought, but I ended up writing some code where I get an `AttributeError` stating that `PS5000a` does not have an attribute named `handle`, which means at some point, the garbage collector has destroyed the attribute, but can still call the `__del__` method.

Reply:
No, it's already defined in the `__init__`. After some looking up, I conclude that it is linked to the behaviour of `__del__`, where the simple attributes are already destroyed when arriving in the `__del__` method. To quote the `__del__` doc :

> It is not guaranteed that [__del__()](https://docs.python.org/3/reference/datamodel.html#object.__del__) methods are called for objects that still exist when the interpreter exits.

After some extra reading, it turns out that `__del__` is at best a makeshift, at worse completely useless, it's better to call the `close()` method anyway.